5 add config file support

senofsky opened the pull request:
I added config file support using toml and marshmallow.

A default config file can be created using `transient init` and `transient init -name config-file-name`.

The default config explains the format. QEMU args remain an arbitrary list of strings.

I added some very basic behave tests, but I still want to add actual unit tests.

I look forward to your feedback and ways we can improve this pull request. Thanks Adam!

senofsky committed:
Add config file tests

I created a few tests that exercise the use of config files.
To do this, I created different config files that are preconfigured
to imitate pre-existing tests.

senofsky committed:
Fix the incorrect Transient-run -name help string

senofsky committed:
Add configuration.py to support configurations

I added configuration.py to provide:

    * Configuration schemas for run, list, and delete
    * Support the use of a configuration file with CLI arguments

I used toml to read TOML-based configuration files and marshmallow
to validate and deserialize Transient configurations.

I added marshmallow and toml to setup.py dependencies.

senofsky committed:
Update cli.py & transient.py to use configuration

I updated cli.py and transient.py to use the features provided
by configuration.py. This includes:

    * adding a '-config' file for selecting a configuration file
    * setting Click options to default to a 'NotGiven'
    * using CLI arguments to create a Config object
    * using the Config class as an argument to Transient

ALSchwalm commented on transient/cli.py:
```diff
@@ -4,6 +4,7 @@
 import signal
 import sys
 
+from . import config as _config
```
I assume this is to avoid the module getting shadowed by a local 'config' variable? I'd rather change the module to 'configuration' or something if this is a common issue.

senofsky replied:
I debated this a bit. Your assumption is correct; it's used to prevent shadowing from local config variables.

I could name the module `configuration.py` but it's asymmetric with `transient/utils.py` and `config/`.

Currently, this module is only used in `cli.py` and `transient.py`.

I don't have a predilection to either - I just wanted to try to keep symmetry with the existing codebase.

senofsky replied:
I needed to add more public members to `config.py` so I ended up renaming this file to `configurations` so `_config` isn't in multiple places.

ALSchwalm commented on transient/cli.py:
```diff
@@ -281,6 +284,24 @@ def _find_requested_images(
     return images
 
 
+@click.help_option("-h", "--help")
+@click.option("-name", help="Creates a config file with the given name")
+@cli_entry.command("init")
+def init_impl(**kwargs: Any) -> None:
+    """Creates a default Transient config file """
+    args = argparse.Namespace(**kwargs)
+
+    if args.name:
+        config_file = args.name
+    else:
+        config_file = "transient-config.toml"
+
+    with open(config_file, "w") as file:
+        file.write(_config.default_config_file)
+
+    sys.exit(0)
```
I think I'm opposed to this 'init' functionality. In my mind, `init` is typically associated with a command that has some special effect on the given directory. `git init` make the directory a git repo, `vagrant init` makes it a vagrant dir, `cargo init`, etc. But transient, by design, has no special effect on the directory. There are no hidden state directories there.

We could just change the name, but then we also have the problem of the default config not having any meaningful default image. The usage of `centos/7:2004.01` seems arbitrary. I think I'd rather just move the 'default' config in to examples in the docs instead.

senofsky replied:
I shamelessly stole the concept from Vagrant - `vagrant init` initializes the current directory with a Vagrantfile.

It is true that Transient doesn't affect the directory in a special manner.  I'm open to changing the name - the intent of this function is to provide a default configuration file that worked out-of-the-box which developers could easily create and modify:
```
transient init
transient run -config transient-config.toml
```

I would like to push back on keeping a config file example only in docs. During testing, I intend for Transient to create a config and use it, which exercises the config file. Hopefully, this will prevent the config file schema from falling out of date - especially compared to a docs page. However, it would be beneficial to have a single docs page that enumerates all possible config options with examples.

The image I picked was arbitrary.

How would you like to proceed?

ALSchwalm replied:
I agree that I don't want us to get into a position where the docs conflict with the actual schema. I would rather put example configs in a directory and include those files in the docs and the tests. This ensures that _every_ example we give does actually work.

I don't think I am really persuaded by the out-of-the-box concept. I expect a developer to go more in the direction of: 

1. I need to run an ubuntu vm: `transient run -image ubuntu:whaterver -ssh -- -m 1G`
2. Oh, I also need a shared folder `.. -s host:guest`
3. Ah, I guess I also do want to wait for shutdown `... -shutdown-timeout 0`
4. I guess I'll make a config file rather than this long commandline 

Any options we pick for the default file will be arbitrary and not necessarily correct anyway. The usage of `-enable-kvm`, for example, won't work when `/dev/kvm` is not available (e.g., under docker). So, I remain opposed to having `init` functionality like this.

senofsky replied:
That's a well-reasoned argument and I am inclined to agree with you regarding a configuration file driven by need rather than a boilerplate config - especially given the simple interface.

To resolve this, I will:

1. remove the `init` command
2. move the exhaustive config file string to `test/config-files/exhaustive-config`

Let me know if you disagree and/or have any additional comments.

senofsky replied:
Resolved with my proposed solution.

ALSchwalm commented on transient/config.py:
```diff
@@ -0,0 +1,261 @@
+"""
+
+
+class Config(Dict[Any, Any]):
```
Isn't this exactly what argparse.Namespace is? Maybe just make an alias for that instead of this? We aren't really getting anything from the `Dict[Any, Any]` type info anyway.

senofsky replied:
The behavior is the same as far as I can tell, which I intended. I saw that you used `argparse.Namespace` directly in `cli.py` but I couldn't find documentation on how to directly use the class. The official docs indicate that it's only used by `parse_args()`: 
 https://docs.python.org/3.8/library/argparse.html#namespace 

Additionally, I have to combine arguments from the CLI and the configuration file - I'm not sure how I would accomplish that with using `argparse.Namespace` and if there are any side-effects of trying to modify the class directly without using `argparse`. I haven't found any documentation that mentions how to do this. 

Ultimately, `argparse.Namespace` is opaque to me and it seems, to my knowledge, only used by the `argparse` library. If I'm missing something, feel free to point me in the right direction and I can remove this class.

senofsky replied:
I've pushed my latest changes but this is still unresolved. Let me know how you want to move forward with this.

senofsky replied:
I'm going to mark this as resolved. But if you want me to replace this with `argparse.Namespace`, please let me know.

ALSchwalm commented on transient/config.py:
```diff
@@ -0,0 +1,261 @@
+        return Config(**data)
+
+
+def replace_hyphens_with_underscores_in_dict_keys(
```
As we aren't explicitly listing exports, this should start with a `_` to indicate it is local to this module.

senofsky replied:
Prefix added.

ALSchwalm commented on transient/config.py:
```diff
@@ -0,0 +1,261 @@
+        parsed_config_file = toml.loads(config_file)
+    except RuntimeError as error:
+        print(error, file=sys.stderr)
+        sys.exit(1)
```
We should not `sys.exit`, except within `cli.py`. If someone is trying to use transient as a library, this will kill their program.

ALSchwalm replied:
I also don't think we should suppress the exception. There is a general question of how to display errors if the user is coming from the CLI, but that should be handled by catching the exceptions there.

senofsky replied:
Replaced `sys.exit(1)` in `config.py` with `raise` in functions:

```
    load_config_file()
    parse_config_file()
    create_transient_run_config_from_cli_args()
```

That way, these functions can still print useful error messages during exception handling and re-raise the exception.

ALSchwalm commented on transient/config.py:
```diff
@@ -0,0 +1,261 @@
+            if string in line:
+                return line_number
+
+    print("Line not found", file=sys.stderr)
```
Definitely shouldn't print this cryptic message here.

senofsky replied:
Print statement deleted. Technically that print statement should have been moved to the calling function and invoked if if the line number is -1.

Technically this shouldn't happen because the matching string comes directly from the config file itself, but I wasn't sure what to do if the line number is -1 for whatever magical reason.

ALSchwalm commented on transient/config.py:
```diff
@@ -0,0 +1,261 @@
+        print(f"Invalid configuration file, line {line_number}: {error}", file=sys.stderr)
+
+
+def load_config_file(config_file_path: str) -> Config:
```
So this is the public function? Then `parse_config_file` should be prefixed with a "_".

senofsky replied:
Technically, the only function that's invoked by an external object is `create_transient_run_config_from_cli_args()`

Therefore, I have prefixed all other functions in this class with the underscore prefix.

ALSchwalm commented on transient/config.py:
```diff
@@ -0,0 +1,261 @@
+        # Combine the arguments from the CLI and configuration file
+        # CLI arguments take precedence
+        for key in config_file:
+            if not config[key] and config_file[key]:
```
I don't fully follow this. Will it not get confused if `config[key]` exists but is falsey (e.g. if it's 0)

senofsky replied:
I struggled to make this code fragment more readable.

The intended behavior is to assign `config[key]` if it's falsey (None, ( ), [ ], ' ', 0, { }) with a NOT falsey `config_file[key]`.

The purpose of checking if `config[key]` is falsey is to prevent being overwritten by the config file.
The purpose of checking if `config_file[key]` is falsey is to prevent unnecessary falsey assignment.

I was thinking of adding a helper function such as `config_option_is_empty()` and `config_option_is_not_empty()` so:

```python
if config_option_is_empty(config[key]) and config_option_is_not_empty(config_file[key]):
```

Any thoughts? I want to make this as easily comprehensible as possible.

ALSchwalm replied:
I still don't follow. Suppose I set `-shutdown-timeout 0` on the commandline, will this not override that value with something from the config file? I don't understand what the _value_ of the configuration matters, don't we only care about whether an option was specified on the commandline at all? If so, we should always use it, regardless of whether the value the user specified happens to be falsey or not.

senofsky replied:
That is an excellent point - I did not think of a CLI argument that is purposefully given but is falsey.

I am now aware of a different problem. The argument dictionary created by `Click` populates each argument value (user-given value, default value, or I believe an empty tuple). But I cannot use falsey comparisons to check if an argument was given by the user because of the reason you mentioned and Default values will always override the config value.

Is there a way to determine if the argument in the argument dictionary returned by `Click` was set via command line, a default value, or not given? I'll continue to think about this.

ALSchwalm replied:
I'm not sure. We can always just set the 'default' value to a python object we define like NotPassed or something. Then we can compare against that, as a user could not pass that value from the command line.

senofsky replied:
I've tried to prototype the idea of passing 'default' values with a `NotGiven` type. I found out that `Click` type inferences the expected type via the default argument, so I have to use `StringNotGiven`, `IntegerNotGiven`, and `ListNotGiven`.

But there is still a problem. Flags are boolean values and default to `False` if not set, _always_  overwriting the config file value. I think the most intuitive case would be to ignore `False` boolean flags.

I'm still thinking through this. Let me know if you have any thoughts.

senofsky replied:
I resolved this by implementing default values that indicate the option was not set on the CLI.

If a flag is not passed on the CLI, the default `False` value does **not** override the value found in the config file.
If a flag is passed on the CLI, it is `True` and will override the value found in the config file.

ALSchwalm commented on transient/config.py:
```diff
@@ -0,0 +1,261 @@
+
+    """
+
+    def __init__(self, *args: Dict[Any, Any], **kwargs: Dict[Any, Any]) -> None:
```
shouldn't 'args' be a `List[Any]`?

senofsky replied:
Yes. Fixed.

ALSchwalm commented on transient/config.py:
```diff
@@ -0,0 +1,261 @@
+    def __init__(self, *args: Dict[Any, Any], **kwargs: Dict[Any, Any]) -> None:
+        super(Config, self).__init__(*args, **kwargs)
+        for arg in args:
+            if isinstance(arg, dict):
```
We just ignore any non-dict arg?

senofsky replied:
Yes. The intent of this class is to take a dictionary and give it the ability to use dot notation for accessing values.

On second thought, it may be better to use `args: List[Dict[Any, Any]]` in the parameter.

senofsky replied:
Resolved by setting the parameter to `args: List[Dict[Any, Any]]`

ALSchwalm reviewed (changes requested):
Thanks for taking the time to work on this!

ALSchwalm commented:
That sounds reasonable to me

On Thu, Jun 18, 2020, 3:29 PM Mark V Senofsky <notifications@github.com>
wrote:

> *@senofsky* commented on this pull request.
> ------------------------------
>
> In transient/cli.py
> <https://github.com/ALSchwalm/transient/pull/74#discussion_r442484678>:
>
> > +@click.help_option("-h", "--help")
> +@click.option("-name", help="Creates a config file with the given name")
> +@cli_entry.command("init")
> +def init_impl(**kwargs: Any) -> None:
> +    """Creates a default Transient config file """
> +    args = argparse.Namespace(**kwargs)
> +
> +    if args.name:
> +        config_file = args.name
> +    else:
> +        config_file = "transient-config.toml"
> +
> +    with open(config_file, "w") as file:
> +        file.write(_config.default_config_file)
> +
> +    sys.exit(0)
>
> That's a well-reasoned argument and I am inclined to agree with you
> regarding a configuration file driven by need rather than a boilerplate
> config - especially given the simple interface.
>
> To resolve this, I will:
>
>    1. remove the init command
>    2. move the exhaustive config file string to
>    test/config-files/exhaustive-config
>
> Let me know if you disagree and/or have any additional comments.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/ALSchwalm/transient/pull/74#discussion_r442484678>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAIDYP5FIE4VO25A5YFYMXTRXJ2LLANCNFSM4OABRUYA>
> .
>

senofsky committed:
Add exhaustive config file example

This configuration file documents all possible configuration
options and how to format them.

ALSchwalm commented on transient/configuration.py:
```diff
@@ -0,0 +1,331 @@
+    """
+
+    def __init(self, error: RuntimeError) -> None:
+        print(f"Failed to parse configuration file: {error}", file=sys.stderr)
```
You should definitely not be printing things in the `__init__` of an exception

ALSchwalm replied:
To be clear, this shouldn't just be changed to a log. I think the best approach is to do this as a log at the site where the exceptional behavior occurs, and then raise the exception.

senofsky replied:
Understood. One thing I'm not familiar with is Python's exception handling and logging practices. I can have: 

```python
except RuntimeError as error:
    logging.error(f"Failed to parse configuration file: {error}")
    raise ConfigFileParsingError(error) 
```
and just catch it in `cli.py` and simply `sys.exit(1)`.

or I can have:
```python
except RuntimeError as error:
    raise ConfigFileParsingError(error)
```
and catch it in `cli.py`, print the error response, and `sys.exit(1)`.

ALSchwalm replied:
I think I prefer the former. No particular reason though, I suppose if something else wrapped/converted the exception then we still have a record of the original exception? idk

senofsky replied:
Used former solution.

ALSchwalm commented on transient/configuration.py:
```diff
@@ -0,0 +1,331 @@
+from typing import Any, Dict, List, MutableMapping
+
+
+IntegerNotGiven = -1
+StringNotGiven = "StringNotGiven"
+ListNotGiven = [StringNotGiven]
+TupleNotGiven = (StringNotGiven,)
```
I don't think this is quite how this should be done. Rather than having magic values to indicate that a user has not passed the argument, we should set the default to `None` (or rather, just not set a default for anything in the cli.py file). The user _can't_ pass `None` for any flag, so this unambiguously means that a value has not been passed. The current approach will incorrectly believe the user has not passed a flag if they happen to specify `-1`, for example, which is a valid integer.

senofsky replied:
Removing all defaults and setting them in marshmallow's schema was definitely the cleanest and easiest way to do this. I wish I thought of it earlier, thanks.

The only minor caveat is that the default is `None` for most options, but the ones that have `multiple=True` are set to empty tuples and flags are still False.

ALSchwalm commented on transient/configuration.py:
```diff
@@ -0,0 +1,331 @@
+    return True
+
+
+def _option_was_not_set_in_cli(option: Any) -> bool:
+    """Returns True if an option was NOT set in the command line
+    """
+    return not _option_was_set_in_cli(option)
```
I think defining this is non-idiomatic

senofsky replied:
This is non-idiomatic, but I tend to write code in a way that I _think_ is more readable and understandable.

In my _personal opinion_, I believe:
```
if _option_was_not_set_in_cli(option):
    do_stuff()
```
is more readable and understandable than:
```
if not option_was_set_in_cli(option):
   do_stuff()
```

but this isn't a hill that I would choose to die on. If this is something you disagree with, let me know and I'll change it.

ALSchwalm replied:
Yes, I would prefer it removed

senofsky replied:
Function deleted.

ALSchwalm commented on transient/configuration.py:
```diff
@@ -0,0 +1,331 @@
+        )
+
+        if line_number != -1:
+            print(
+                f"Invalid option on line {line_number} in configuration file: {invalid_option}",
+                file=sys.stderr,
+            )
```
Any print to sys.stderr should instead be a `logging.error`

senofsky replied:
I just copied existing logic in `cli.py`:

```
print("No images match selection", file=sys.stderr)
```

Should I change these to use `logging.error` as well?

ALSchwalm replied:
No, cli.py can print to stderr, as it would only be used by someone using the command line. Things outside of that module should be written as though they could be used from a lib, so avoid writing to stdout/stderr when possible

senofsky replied:
Understood. I implemented your request.

ALSchwalm commented on transient/configuration.py:
```diff
@@ -0,0 +1,331 @@
+    shared_folder = fields.List(fields.Str(), missing=[])
+
+
+def _option_was_set_in_cli(option: Any) -> bool:
+    """Returns True if an option was set in the command line
+    """
+    if (
+        option == IntegerNotGiven
+        or option == StringNotGiven
+        or option == ListNotGiven
+        or option == TupleNotGiven
+        or option == False
+    ):
+        return False
+
+    return True
```
Following my above comment, this is reduced to just `option is not None`

senofsky replied:
After implementing the recommended changes to have `cli.py` not set defaults (and instead, defaults are set via marshmallow schema) this is reduced to:

```
if option is None or option == () or option is False:
```

If a `Click` option contains `multiple=True`, the default is an empty tuple. And as previously mentioned, flags default to False.

ALSchwalm commented on transient/configuration.py:
```diff
@@ -0,0 +1,331 @@
+        self.__setitem__(key, value)
+
+    def __setitem__(self, key: Any, value: Any) -> None:
+        super(Config, self).__setitem__(key, value)
+        self.__dict__.update({key: value})
```
I don't understand why both of these are necessary (or indeed, why we need to define `__setitem__` at all). We inherit from Dict, so the default behavior should be sufficient, shouldn't it? A trivial example seems to work fine:

```
>>> class Example(dict):
...     def __getattr__(self, key):
...         return self.get(key)
...     def __setattr__(self, key, val):
...         self[key] = val
... 
>>> e = Example()
>>> e.foo = 10
>>> e.foo
10
>>> e["foo"]
10
>>>
```

senofsky replied:
You're right. Code fragment deleted.

ALSchwalm commented on transient/configuration.py:
```diff
@@ -0,0 +1,331 @@
+        return Config(**data)
+
+
+class _TransientDeleteConfigSchema(_TransientConfigSchema):
+    """Defines the schema for the Transient-delete configuration and validates
+       the fields during deserialization
+    """
+
+    force = fields.Bool(missing=False)
+
+
+class _TransientListConfigSchema(_TransientConfigSchema):
+    """Defines the schema for the Transient-list configuration and validates
+       the fields during deserialization
+
+       Note that this class is a wrapper to maintain symmetry with the other
+       schemas.
+    """
+
+    pass
```
I'm not sure that allowing the config stuff for list/delete is wise. I would at least like to see some tests for it. The issue I think of here is that there are slightly different meanings to the usage of `-name` and `-image` between deleteing and running a VM. For example:

Suppose you have a config that specifies a image but no name, so it creates a VM with a UUID name:

```
[transient]
image = ["centos/7:2004.01"]
```

Then you run `transient delete -c /the/above/config`, you will actually delete the backend image "centos/7:2004.01", because that is the behavior of `transient delete -image centos/7:2004.01` (deleting an image with no name deletes the backend image and all associated frontend images that were backed by that image).

That said, it could just be considered a user error (and they will still get the warning), so I'm probably Ok with this, it should just have tests for list/delete with a config file.

senofsky replied:
I agree that using a configuration file for list or delete would be a bad idea. I think it would be too easy to accidentally execute `transient delete` with a given config file, causing a negative user experience.

That's why I did not implement that functionality. Using a configuration file is only possible for `Transient-run`. The reason for multiple schema classes is to have a symmetric configuration schemas between `run`, `list`, and `delete`.

Essentially,  all command line arguments are passed to marshmallow for initialization (setting default values if required) and validation. Then, marshmallow returns the validated Config object for use. This happens for `run`, `list`, and `delete`.

ALSchwalm reviewed (changes requested):
Thanks for addressing my previous concerns, I did another pass and have a few other concerns.

Thanks for sticking with this!

ALSchwalm commented on transient/configuration.py:
```diff
@@ -0,0 +1,297 @@
+    def __delattr__(self, item: Any) -> None:
+        self.__delitem__(item)
+
+    def __delitem__(self, key: Any) -> None:
+        super(Config, self).__delitem__(key)
+        del self.__dict__[key]
```
Same comment as before, why do we need to define `__delitem__`? I assume the default behavior would be fine.

ALSchwalm reviewed (changes requested):
One minor comment, but otherwise, this looks good!

senofsky commented:
I definitely learned a few things while creating this PR so thanks for the opportunity!

Let me know if there is anything else I can clean up.

ALSchwalm commented:
lgtm :+1:
